### PR TITLE
Remove frame-local variables usage

### PR DESCRIPTION
**Pull request by john2x**
Hi. Now that Emacs 26.1 has been released, I thought it would be a good time to open a PR for my changes :) I've been using Emacs 26 pre-releases for a few months now (plus a few months more on Emacs 25) with my changes on perspective-el, and I haven't run into any issues.
(#64)

### CODE REVIEW

File: perspective.el [hunk 1]
```diff
@@ -196,54 +166,54 @@ Run with the activated perspective active.")
...
-;; but I don't want to have to manually munge frame-parameters
-;; all the time so I'm using it anyway.
-(make-variable-frame-local
- (defvar perspectives-hash nil
-   "A hash containing all perspectives. The keys are the
-perspectives' names. The values are persp structs,
-with the fields NAME, WINDOW-CONFIGURATION, BUFFERS,
-BUFFER-HISTORY, KILLED, POINT-MARKER, and LOCAL-VARIABLES.
+;; Keeping the variable definitions here since the docstrings might
+;; still be useful for developers.
```
Comment (nex3) [inline perspective.el]: Just get rid of these; this is why Git history exists.

File: perspective.el [hunk 2]
```diff
@@ -293,9 +263,9 @@ for the perspective."
       (dotimes (_ 2) (push (pop args) keywords)))
     (setq keywords (reverse keywords))
     `(let ((persp (make-persp-internal ,@keywords)))
-       (when persp-curr
-         (setf (persp-local-variables persp) (persp-local-variables persp-curr)))
-       (puthash (persp-name persp) persp perspectives-hash)
+       (when (frame-parameter nil 'persp-curr)
```
Comment (nex3) [inline perspective.el]: Wrap `(frame-parameter nil 'persp-curr)` in a function so you don't have to repeat it all over the place. Same with other frequently-accessed frame parameters.

File: perspective.el [hunk 3]
```diff
@@ -408,23 +378,28 @@ EVENT is the click event triggering this function call."
   (interactive "e")
   (persp-switch (format "%s" (car (posn-string (event-start event))))))
 
+(defun persp-mode-line ()
+  "Use this in the mode-line to show up-to-date value of `persp-modestring'."
```
Comment (nex3) [inline perspective.el]: `persp-modestring` doesn't refer to anything anymore. I'd write this docstring as "Return the string displayed in the modeline representing the perspectives."

Similarly, update other docstrings that refer to `persp-modestring`.

File: perspective.el [hunk 4]
```diff
@@ -408,23 +378,28 @@ EVENT is the click event triggering this function call."
...
           (close (list (nth 1 persp-modestring-dividers)))
           (sep (nth 2 persp-modestring-dividers)))
-     (setq persp-modestring
+      (set-frame-parameter nil 'persp-modestring
            (append open
                    (persp-intersperse (mapcar 'persp-format-name
                                               (persp-names)) sep)
                    close)))))
 
+
```
Comment (nex3) [inline perspective.el]: Extra newline

File: perspective.el [hunk 5]
```diff
@@ -618,35 +593,36 @@ See also `persp-switch' and `persp-add-buffer'."
...
-  (when (equal name (persp-name persp-last))
-    (setq persp-last nil))
-  (when (equal name (persp-name persp-curr))
+  (when (equal name (persp-name (frame-parameter nil 'persp-last)))
+    (set-frame-parameter nil 'persp-last nil))
+  (when (equal name (persp-name (frame-parameter nil 'persp-curr)))
+    (persp-switch (persp-find-some))
     ;; Don't let persp-last get set to the deleted persp.
-    (persp-frame-local-let ((persp-last persp-last)) (persp-switch (persp-find-some)))))
+    (set-frame-parameter nil 'persp-last (frame-parameter nil 'nil))))
```
Comment (nex3) [inline perspective.el]: I'm worried that setting `persp-last` to `nil` will break existing behavior. Can you set it to `(persp-find-some)` in its frame instead, as before?

File: perspective.el [hunk 6]
```diff
@@ -781,22 +757,22 @@ See also `persp-add-buffer'."
...
-        ad-do-it
-        ;; We want the buffer lists that were created in the recursive edit,
-        ;; but not the window configurations
-        (maphash (lambda (key new-persp)
-                   (let ((persp (gethash key old-hash)))
-                     (when persp
-                       (setf (persp-buffers persp) (persp-buffers new-persp)))))
-                 perspectives-hash)
-        (setq perspectives-hash old-hash)))))
+    (set-frame-parameter nil 'persp-recursive (frame-parameter nil 'persp-curr))
```
Comment (nex3) [inline perspective.el]: Replacing `let` with `set` doesn't preserve the existing behavior--the frame parameter won't be reset after the function is called, or if it produces an exception.

I suggest replacing `persp-frame-local-let` with a new `persp-let-frame-parameter` macro that sets frame parameters for the duration of the body, and gracefully handles errors like `persp-frame-local-let`.

File: perspective.el [hunk 7]
```diff
@@ -817,15 +793,18 @@ named collections of buffers and window configurations."
...
     (remove-hook 'after-make-frame-functions 'persp-init-frame)
     (remove-hook 'ido-make-buffer-list-hook 'persp-set-ido-buffers)
     (setq read-buffer-function nil)
-    (setq perspectives-hash nil)
-    (setq global-mode-string (delq 'persp-modestring global-mode-string))))
+    (set-frame-parameter nil 'perspectives-hash nil)
+    (setq global-mode-string (delete '(:eval (persp-mode-line)) global-mode-string))
+    (set-default 'header-line-format (delete '(:eval (persp-mode-line)) header-line-format))
+    (unless (delete "" header-line-format)
+      (set-default 'header-line-format nil))))
```
Comment (nex3) [inline perspective.el]: I don't understand what this statement is doing. Can you add a comment explaining?

File: perspective.el (hunk 1 above)
Comment (john2x) [reply perspective.el]: I've removed the commented out variables.

File: perspective.el (hunk 2 above)
Comment (john2x) [reply perspective.el]: I've added wrapper functions for getting `persp-curr` and `persp-last`.

File: perspective.el (hunk 3 above)
Comment (john2x) [reply perspective.el]: I've updated the docstrings refering to `persp-modestring`,

File: perspective.el (hunk 5 above)
Comment (john2x) [reply perspective.el]: I did some improvements around this, but I am not sure if it's the correct approach.

File: perspective.el (hunk 6 above)
Comment (john2x) [reply perspective.el]: Added a new macro `persp-let-frame-parameters`, inspired by `persp-frame-local-let`.

File: perspective.el (hunk 7 above)
Comment (john2x) [reply perspective.el]: Added a comment to this line. It seems that setting `header-line-format` to an empty string still renders it in the buffer. Setting it to nil removes it.

**Comment by john2x** [comment]
Thanks for the review! I've pushed some changes to address the inline comments.

File: perspective.el [hunk 8]
```diff
@@ -161,60 +173,19 @@ Run with the activated perspective active.")
 (define-key perspective-map (kbd "<left>") 'persp-prev)
 (define-key perspective-map persp-mode-prefix-key 'persp-switch-last)
 
+(defun persp--current-persp ()
```
Comment (nex3) [inline perspective.el]: Please make this and `persp-last` public. We want users to have access to this information.

Comment (john2x) [reply perspective.el]: Changed `persp--current-persp` and `persp--last-persp` to `persp-current-persp` and `persp-last-persp`.

**Pull request merged**